lib/spack/external/defer_signals.py

```diff
@@ -0,0 +1,97 @@
+# From https://gist.github.com/evansd/2375136
+#
+# The MIT License (MIT)
+#
+# Copyright (c) 2013 David Evans
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of
+# this software and associated documentation files (the "Software"), to deal in
+# the Software without restriction, including without limitation the rights to
+# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+# the Software, and to permit persons to whom the Software is furnished to do so,
+# subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+# FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+# IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import os
+import signal
+import six
+
+
+class defer_signals(object):
+    """
+    Context manager to defer signal handling until context exits.
+
+    Takes optional list of signals to defer (default: SIGHUP, SIGINT, SIGTERM).
+    Signals can be identified by number or by name.
+
+    Allows you to wrap instruction sequences that ought to be atomic and ensure
```

Comment:
I don't think this is possible: the OS can context switch without signals.

Reply:
I agree but this is an external library, and we do not actually care if we context-switch here -- we only care about the main thread in the daemon process -- the one that receives signals.

Reply:
I see what you mean and that's pretty cool, but to make sure, does is the signal *required* to actually change the state of the process? For example, where this is used, there is some code:

```
        with defer_signals(...):
             bg = self._is_background()
             # <--- say we switch from background to foreground here
             if not bg and any(flags):    # fg, but input not enabled
```

My understanding is that the process can switch between the BG and FG in between collecting the state and checking the state. The signal might not be received (or rather the handler might be delayed), but the state of the process has changed.

So I think this logic may be necessary but I'm not sure if it's sufficient.